### ISSUE TITLE
Better documentation

### ISSUE THREAD

**Issue by gdalle** [axelparmentier/InferOpt.jl#70]
We urgently need a better documentation with
- a better intro to the theory
- examples for all types of loss functions
- examples for using kwargs to pass instances (#56)
- a guide for choosing your layer [_à la scikit-learn_](https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html)

**Comment by gdalle** [comment]
I also wanna write an explanation of the design of the library:
- there are layers and losses, all callable
- some layers are probabilistic, others not, but all are differentiable
- most losses must be built with a layer
- if you have your own differentiable loss you can compose it with a layer
- if you have your own non-differentiable loss you can pushforward it with a probabilistic layer
